File: Code/Laravel-Backend/app/Http/Controllers/BillingController.php
<?php

namespace App\Http\Controllers;

use App\Exceptions\BillingValidationException;
use App\Exceptions\InvoiceNotFoundException;
use App\Exceptions\ProductNotFoundException;
use App\Http\Resources\InvoiceResource;
use App\Http\Resources\ProductResource;
use App\Services\BillingService;
use Illuminate\Http\Request;

class BillingController extends Controller
{
    public function __construct(private BillingService $billingService)
    {
    }

    public function getProduct(string $code)
    {
        try {
            return new ProductResource($this->billingService->getProduct($code));
        } catch (ProductNotFoundException $e) {
            return response()->json(['error' => $e->getMessage()], 404);
        }
    }

    public function createInvoice(Request $request)
    {
        try {
            $invoice = $this->billingService->createInvoice(
                $request->input('customerId'),
                $request->input('items', [])
            );
            return (new InvoiceResource($invoice))->response()->setStatusCode(201);
        } catch (ProductNotFoundException $e) {
            return response()->json(['error' => $e->getMessage()], 404);
        } catch (BillingValidationException $e) {
            return response()->json(['error' => $e->getMessage()], 400);
        }
    }

    public function getInvoice(string $id)
    {
        try {
            return new InvoiceResource($this->billingService->getInvoice($id));
        } catch (InvoiceNotFoundException $e) {
            return response()->json(['error' => $e->getMessage()], 404);
        }
    }

    public function cancelInvoice(string $id)
    {
        try {
            return new InvoiceResource($this->billingService->cancelInvoice($id));
        } catch (InvoiceNotFoundException $e) {
            return response()->json(['error' => $e->getMessage()], 404);
        }
    }

    public function listInvoices(Request $request)
    {
        return InvoiceResource::collection(
            $this->billingService->listInvoices($request->query('date'))
        );
    }
}

File: Code/DotNet-Backend/Controllers/BillingController.cs
using BillingApi.Models;
using BillingApi.Services;
using Microsoft.AspNetCore.Mvc;

namespace BillingApi.Controllers;

[ApiController]
[Route("api")]
public class BillingController : ControllerBase
{
    private readonly BillingService _billingService;

    public BillingController(BillingService billingService)
    {
        _billingService = billingService;
    }

    [HttpGet("products/{code}")]
    public async Task<IActionResult> GetProduct(string code)
    {
        try
        {
            return Ok(await _billingService.GetProductAsync(code));
        }
        catch (ProductNotFoundException e)
        {
            return NotFound(new { error = e.Message });
        }
    }

    [HttpPost("invoices")]
    public async Task<IActionResult> CreateInvoice([FromBody] CreateInvoiceRequest request)
    {
        try
        {
            var invoice = await _billingService.CreateInvoiceAsync(request);
            return StatusCode(201, invoice);
        }
        catch (ProductNotFoundException e)
        {
            return NotFound(new { error = e.Message });
        }
        catch (BillingValidationException e)
        {
            return BadRequest(new { error = e.Message });
        }
    }

    [HttpGet("invoices/{id}")]
    public async Task<IActionResult> GetInvoice(string id)
    {
        if (!Guid.TryParse(id, out var guid))
        {
            return NotFound(new { error = $"Invoice not found: {id}" });
        }
        try
        {
            return Ok(await _billingService.GetInvoiceAsync(guid));
        }
        catch (InvoiceNotFoundException e)
        {
            return NotFound(new { error = e.Message });
        }
    }

    [HttpPost("invoices/{id}/cancel")]
    public async Task<IActionResult> CancelInvoice(string id)
    {
        if (!Guid.TryParse(id, out var guid))
        {
            return NotFound(new { error = $"Invoice not found: {id}" });
        }
        try
        {
            return Ok(await _billingService.CancelInvoiceAsync(guid));
        }
        catch (InvoiceNotFoundException e)
        {
            return NotFound(new { error = e.Message });
        }
    }

    [HttpGet("invoices")]
    public async Task<IActionResult> ListInvoices([FromQuery] string? date)
    {
        return Ok(await _billingService.ListInvoicesAsync(date));
    }
}
